onnxruntime/test/onnx/callables.h

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <type_traits>
+#include <tuple>
+
+namespace onnxruntime {
+namespace test {
+
+/// <summary>
+/// This class serves as a callable entity that can be used
+/// either in a threadpool or as a callback for async operations. It supports both
+/// plain functions as well as member functions and provides type
+/// erasure functionality. Unlike std::function, it
+/// does NOT do any memory allocations and it is cheap to copy.
+/// </summary>
+/// 
+/// <example>
+/// Plain function:
+///   int g(void* context, float);
+///   Callable<int, float> cb(nullptr, g);
+///   int result = cb.Invoke(1.2f);
+///
+/// Class method:
+/// class A {
+///    int f(float);
+/// };
+/// 
+/// A a;
+/// CallableFactory<A, int, float> f(&a);
+/// - We get back Callable<int, float> as above but call instance method
+/// auto cb = f.GetCallable<&A::f>();
+/// int result = cb.Invoke(1.2f);
+/// </example>
+/// 
+/// <typeparam name="Result">Call back result type</typeparam>
+/// <typeparam name="...Args">Types of args for the callable</typeparam>
+template <typename    Result,
+          typename... Args>
+class Callable {
+public:
+
+  using
+  ContextType = void;
+
+  using
+  ResultType = Result;
+
+  using
+  FunctionType = ResultType(*)(ContextType*, Args...);
+
+  // You can use this directly if you have a simple
+  // function like C-function or a stateless lambda
+  // For methods we want to make use of AsyncMethodCallback
+  // And covert it to a Callable
+  Callable(ContextType*  context,
+    FunctionType func) :
+    context_(context),
+    func_(func) {
+  }
+
+  Callable() : Callable(nullptr, nullptr) {}
+
+  Callable(const Callable&) = default;
+  Callable& operator=(const Callable&) = default;
+
+  Callable(Callable&&) = default;
+  Callable& operator=(Callable&&) = default;
+
+  void Clear() {
+    context_ = nullptr;
+    func_ = nullptr;
+  }
+
+  operator bool() const { return func_ != nullptr; }
+
+  bool Valid() const { return func_ != nullptr; }
+
+  ContextType* GetContext() const { return context_; }
+
+  FunctionType GetFunctionPtr() const { return func_; }
+
+  // Differentiate instantiation on the presence
+  // of return value. The below two overloads
+  // must be mutually exclusive
+  template<typename T = ResultType>
+  typename std::enable_if<!std::is_same<void, T>::value, T>::type
+  Invoke(Args... args) const {
+    return func_(context_, std::forward<Args>(args)...);
+  }
+
+  template <typename T = ResultType>
+  typename std::enable_if<std::is_same<void, T>::value, void>::type
+  Invoke(Args... args) const {
```

Comment:
>Invoke [](start = 2, length = 6)

Would it be more compatible to have an operator () (Args ... args) here so looks from consumer it behaves rightly like a function?

Reply:
Compatible with what? I prefer an explicit call here.

---
In reply to: [490623318](https://github.com/microsoft/onnxruntime/pull/5169#discussion_r490623318) [](ancestors = 490623318)

Reply:
Meaning the calling of callable could thereby be more like a function call:

func(...)

So if a callable passed into another context which also accept std functions or lambdas, the invoking grammar stays the same to be of an operator (...) format.
I understand as of now there may be no such case or occasion at all, but for scalability's sake why not we do it one step further?

---
In reply to: [491118553](https://github.com/microsoft/onnxruntime/pull/5169#discussion_r491118553) [](ancestors = 491118553,490623318)

Reply:
I am reluctant to do this at this time. I see no value in it.

---
In reply to: [491142725](https://github.com/microsoft/onnxruntime/pull/5169#discussion_r491142725) [](ancestors = 491142725,491118553,490623318)